yew/src/html/listener/registry.rs

```diff
@@ -0,0 +1,503 @@
+use crate::virtual_dom::{Listener, Listeners};
+use std::{
+    cell::RefCell,
+    collections::{HashMap, HashSet},
+    rc::Rc,
+};
+use wasm_bindgen::{prelude::*, JsCast};
+use web_sys::{Element, Event};
+
+thread_local! {
+    /// Global event listener registry
+    static REGISTRY: RefCell<Registry> = Default::default();
+
+    /// Key used to store listener id on element
+    static LISTENER_ID_PROP: wasm_bindgen::JsValue = "__yew_listener_id".into();
+
+    static BODY: web_sys::HtmlElement = web_sys::window()
+        .expect("no window global")
+        .document()
+        .expect("no document on window")
+        .body()
+        .expect("no body on document")
+}
+
+#[derive(Clone, Copy, std::hash::Hash, Eq, PartialEq, Debug)]
+struct EventDescriptor {
+    kind: &'static str,
+    flags: u8,
+}
+
+impl From<&dyn Listener> for EventDescriptor {
+    fn from(l: &dyn Listener) -> Self {
+        Self {
+            kind: l.kind(),
+            flags: l.flags(),
+        }
+    }
+}
+
+/// Global multiplexing event handler registry
+#[derive(Default, Debug)]
+struct Registry {
+    /// Counter for assigning new IDs
+    id_counter: u64,
+
+    /// Events with registered handlers that are possibly passive
+    handling: HashSet<EventDescriptor>,
+
+    /// Events that have been registered as bubbling at least once
+    bubbling: HashSet<EventDescriptor>,
+
+    /// Contains all registered event listeners by listener ID
+    by_id: HashMap<u64, HashMap<EventDescriptor, Rc<dyn Listener>>>,
+
+    /// Keep track of all listeners to drop them on registry drop.
+    /// The registry is never dropped in production.
+    #[cfg(test)]
+    #[allow(clippy::type_complexity)]
+    registered: Vec<(&'static str, Closure<dyn std::ops::Fn(web_sys::Event)>)>,
```

Comment:
I'm with Clippy on this one, it wouldn't hurt to have a type alias for `Closure<dyn std::ops::Fn(web_sys::Event)>`

Reply:
I'd argue that a test-only type alias is more complexity.

Reply:
I didn't notice it was only enabled for tests. Makes sense to ignore the warning then but I have to ask, is it really necessary to have so much test-only code for the registry?

Reply:
I need to unregister existing listeners to start from a clean slate on each test. This is not required in production because listeners are never cleared.

Reply:
I understand why you're doing it, I'm wondering whether there's a way around it. I'm sure you agree that having this extra code for tests isn't the most elegant solution.

Reply:
I find it better than having code in prod that only runs in tests. Matter of picking your poison.